app/screens/settings-screen/username-screen.tsx

```diff
@@ -62,24 +62,27 @@ export const UsernameScreen = ({ navigation }: Props) => {
     }
   `)
 
-  React.useEffect(() => {
-    const isValid = assertValidInput()
-
-    if (!isValid) {
+  useEffect(() => {
+    if (!assertValidInput()) {
       return
     }
 
     usernameExistsQuery({ variables: { username: input } })
   }, [input])
 
-  React.useEffect(() => {
-    if (init && input.length <= 2) {
+  useEffect(() => {
+    if (!assertValidInputLength()) {
+      if (shouldShowErrorMessage) {
+        setMessage(translate("UsernameScreen.3CharactersMinimum"))
+        setMessageIsError(true)
+      }
+      inputForm.current.focus()
       return
     }
 
-    const isValid = assertValidInput()
-
-    if (!isValid) {
+    if (!assertValidInputCharacters()) {
+      setMessage(translate("UsernameScreen.letterAndNumber"))
```

Comment:
Instead of showing an error message can't we just disallow all non-accepted characters at the input?
So if I type:
```
abc%12e
```
the input box would just not accept `%` as input and read:
```
abc12e
```

Reply:
The one problem with this approach is a [limitation of React Native.](https://github.com/facebook/react-native/issues/28535#issuecomment-611196270) The text will flicker when adding one of the disallowed characters. Say my input is "Justin" and i type '?'. The screen will briefly show "Justin?" before rendering back to "Justin".

Reply:
I see, didn't know that about react-native. In web it works fine.

But is a Flickr worse than a warning? Is there another way to improve UX here.... perhaps limit the characters on the keyboard?

Reply:
I looked into the keyboard types available for inputs and none match what we want. I personally prefer the warning, but I it's not by much and subjective.

Reply:
Do what you prefer